### PR TITLE
Remove E0245; improve E0404

### LINKED ISSUE #36337: Erroneously returning E0404 instead of E0245
I was working on #32777 and trying to generate E0245. I came up with the following example:

``` rust
struct Foo(u8);

fn foo<T>(t: T)
    where T: Foo {
}
```

I believe this should return E0245, as `Foo` is not a trait, so it cannot be used as a bound. However, the compiler gives error E404, which states that `Foo` is not a trait, so it cannot be implemented for a type (according to https://doc.rust-lang.org/error-index.html#E0404).

### PR DESCRIPTION
**Pull request by mark-i-m**
Fix #36337

Somehow this is currently breaking --explain, but I don't understand how.

r? @estebank

### CODE REVIEW

File: src/librustc_resolve/diagnostics.rs [hunk 1]
```diff
@@ -1674,6 +1674,29 @@ fn main() {
 ```
 "##,
 
+E0910: r##"
+You tried to use something as a trait bound that is not a trait.
```
Comment (estebank) [inline src/librustc_resolve/diagnostics.rs]: You tried to use something that is not a trait as a trait bound.

File: src/librustc_resolve/diagnostics.rs [hunk 2]
```diff
@@ -1674,6 +1674,29 @@ fn main() {
 ```
 "##,
 
+E0910: r##"
```
Comment (estebank) [inline src/librustc_resolve/diagnostics.rs]: Is 909 the biggest error code so far? I could have sworn we're still <700.

Comment (mark-i-m) [reply src/librustc_resolve/diagnostics.rs]: Oh, I'm off by two... The largest I found was 907: https://github.com/rust-lang/rust/blob/b1f8e6fb06d7362eeb2065347a7db94e76b1cb2f/src/librustc_typeck/diagnostics.rs#L4829

I wasn't really sure how to choose the error number, though...

File: src/librustc_resolve/diagnostics.rs (hunk 1 above)
Comment (mark-i-m) [reply src/librustc_resolve/diagnostics.rs]: Fixed

File: src/librustc_resolve/diagnostics.rs (hunk 2 above)
Comment (estebank) [reply src/librustc_resolve/diagnostics.rs]: Tidy is supposed to tell you:
```
[00:05:36] * 534 error codes
[00:05:36] * highest error code: E0908
[00:05:36] * 175 features
```

I think that somebody sneaked a high error code skipping a bunch somewhere 👀

**Comment by estebank** [comment]
r=me once travis passes

**Comment by mark-i-m** [comment]
@estebank There are ui tests failing for me locally...

These are the test failures I am currently getting that I cannot understand:

<details>

```
failures:

---- [ui] ui/explain.rs stdout ----
	
error: test compilation failed although it shouldn't!
status: exit code: 101
command: "/nobackup/rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/nobackup/rust/src/test/ui/explain.rs" "-L" "/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui/explain.stage1-x86_64-unknown-linux-gnu" "-Crpath" "-O" "-Zmiri" "-Zunstable-options" "-Lnative=/nobackup/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--explain" "E0591" "-L" "/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui/explain.stage1-x86_64-unknown-linux-gnu.aux" "-A" "unused"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
{"message":"no extended information for E0591","code":null,"level":"error","spans":[],"children":[],"rendered":"error: no extended information for E0591\n\n"}

------------------------------------------

thread '[ui] ui/explain.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:2892:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- [ui] ui/lint/use_suggestion_json.rs stdout ----
	diff of stderr:

2	  "message": "cannot find type `Iter` in this scope",
3	  "code": {
4	    "code": "E0412",
-	    "explanation": "
-	The type name used is not in scope.
-	
-	Erroneous code examples:
-	
-	```compile_fail,E0412
-	impl Something {} // error: type name `Something` is not in scope
-	
-	// or:
-	
-	trait Foo {
-	    fn bar(N); // error: type name `N` is not in scope
-	}
-	
-	// or:
-	
-	fn foo(x: T) {} // type name `T` is not in scope
-	```
-	
-	To fix this error, please verify you didn't misspell the type name, you did
-	declare it or imported it into the scope. Examples:
-	
-	```
-	struct Something;
-	
-	impl Something {} // ok!
-	
-	// or:
-	
-	trait Foo {
-	    type N;
-	
-	    fn bar(_: Self::N); // ok!
-	}
-	
-	// or:
-	
-	fn foo<T>(x: T) {} // ok!
-	```
-	
-	Another case that causes this error is when a type is imported into a parent
-	module. To fix this, you can follow the suggestion and use File directly or
-	`use super::File;` which will import the types from the parent namespace. An
-	example that causes this error is below:
-	
-	```compile_fail,E0412
-	use std::fs::File;
-	
-	mod foo {
-	    fn some_function(f: File) {}
-	}
-	```
-	
-	```
-	use std::fs::File;
-	
-	mod foo {
-	    // either
-	    use super::File;
-	    // or
-	    // use std::fs::File;
-	    fn foo(f: File) {}
-	}
-	# fn main() {} // don't insert it for us; that'll break imports
-	```
-	"
+	    "explanation": null
71	  },
72	  "level": "error",
73	  "spans": [

89	      ],
90	      "label": "not found in this scope",
91	      "suggested_replacement": null,
+	      "suggestion_approximate": null,
92	      "expansion": null
93	    }
94	  ],

118	          "suggested_replacement": "use std::collections::binary_heap::Iter;
119	
120	",
+	          "suggestion_approximate": null,
121	          "expansion": null
122	        },
123	        {

140	          "suggested_replacement": "use std::collections::btree_map::Iter;
141	
142	",
+	          "suggestion_approximate": null,
143	          "expansion": null
144	        },
145	        {

162	          "suggested_replacement": "use std::collections::btree_set::Iter;
163	
164	",
+	          "suggestion_approximate": null,
165	          "expansion": null
166	        },
167	        {

184	          "suggested_replacement": "use std::collections::hash_map::Iter;
185	
186	",
+	          "suggestion_approximate": null,
187	          "expansion": null
188	        },
189	        {

206	          "suggested_replacement": "use std::collections::hash_set::Iter;
207	
208	",
+	          "suggestion_approximate": null,
209	          "expansion": null
210	        },
211	        {

228	          "suggested_replacement": "use std::collections::linked_list::Iter;
229	
230	",
+	          "suggestion_approximate": null,
231	          "expansion": null
232	        },
233	        {

250	          "suggested_replacement": "use std::collections::vec_deque::Iter;
251	
252	",
+	          "suggestion_approximate": null,
253	          "expansion": null
254	        },
255	        {

272	          "suggested_replacement": "use std::option::Iter;
273	
274	",
+	          "suggestion_approximate": null,
275	          "expansion": null
276	        },
277	        {

294	          "suggested_replacement": "use std::path::Iter;
295	
296	",
+	          "suggestion_approximate": null,
297	          "expansion": null
298	        },
299	        {

316	          "suggested_replacement": "use std::result::Iter;
317	
318	",
+	          "suggestion_approximate": null,
319	          "expansion": null
320	        },
321	        {

338	          "suggested_replacement": "use std::slice::Iter;
339	
340	",
+	          "suggestion_approximate": null,
341	          "expansion": null
342	        },
343	        {

360	          "suggested_replacement": "use std::sync::mpsc::Iter;
361	
362	",
+	          "suggestion_approximate": null,
363	          "expansion": null
364	        }
365	      ],


The actual stderr differed from the expected stderr.
Actual stderr saved to /nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui/lint/use_suggestion_json.stderr
To update references, run this command from build directory:
/nobackup/rust/src/test/ui/update-references.sh '/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui' 'lint/use_suggestion_json.rs'

error: 1 errors occurred comparing output.
status: exit code: 101
command: "/nobackup/rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs" "-L" "/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui/lint/use_suggestion_json.stage1-x86_64-unknown-linux-gnu" "-Crpath" "-O" "-Zmiri" "-Zunstable-options" "-Lnative=/nobackup/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--error-format" "pretty-json" "-Zunstable-options" "-L" "/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui/lint/use_suggestion_json.stage1-x86_64-unknown-linux-gnu.aux" "-A" "unused"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
{
  "message": "cannot find type `Iter` in this scope",
  "code": {
    "code": "E0412",
    "explanation": null
  },
  "level": "error",
  "spans": [
    {
      "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
      "byte_start": 907,
      "byte_end": 911,
      "line_start": 21,
      "line_end": 21,
      "column_start": 12,
      "column_end": 16,
      "is_primary": true,
      "text": [
        {
          "text": "    let x: Iter;",
          "highlight_start": 12,
          "highlight_end": 16
        }
      ],
      "label": "not found in this scope",
      "suggested_replacement": null,
      "suggestion_approximate": null,
      "expansion": null
    }
  ],
  "children": [
    {
      "message": "possible candidates are found in other modules, you can import them into scope",
      "code": null,
      "level": "help",
      "spans": [
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::collections::binary_heap::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        },
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::collections::btree_map::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        },
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::collections::btree_set::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        },
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::collections::hash_map::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        },
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::collections::hash_set::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        },
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::collections::linked_list::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        },
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::collections::vec_deque::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        },
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::option::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        },
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::path::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        },
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::result::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        },
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::slice::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        },
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/use_suggestion_json.rs",
          "byte_start": 884,
          "byte_end": 884,
          "line_start": 20,
          "line_end": 20,
          "column_start": 1,
          "column_end": 1,
          "is_primary": true,
          "text": [
            {
              "text": "fn main() {",
              "highlight_start": 1,
              "highlight_end": 1
            }
          ],
          "label": null,
          "suggested_replacement": "use std::sync::mpsc::Iter;\n\n",
          "suggestion_approximate": null,
          "expansion": null
        }
      ],
      "children": [],
      "rendered": null
    }
  ],
  "rendered": "error[E0412]: cannot find type `Iter` in this scope\n  --> /nobackup/rust/src/test/ui/lint/use_suggestion_json.rs:21:12\n   |\n21 |     let x: Iter;\n   |            ^^^^ not found in this scope\nhelp: possible candidates are found in other modules, you can import them into scope\n   |\n20 | use std::collections::binary_heap::Iter;\n   |\n20 | use std::collections::btree_map::Iter;\n   |\n20 | use std::collections::btree_set::Iter;\n   |\n20 | use std::collections::hash_map::Iter;\n   |\nand 8 other candidates\n\n"
}
{
  "message": "aborting due to previous error",
  "code": null,
  "level": "error",
  "spans": [],
  "children": [],
  "rendered": "error: aborting due to previous error\n\n"
}

------------------------------------------

thread '[ui] ui/lint/use_suggestion_json.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:2892:9

---- [ui] ui/lint/unused_parens_json_suggestion.rs stdout ----
	diff of stderr:

24	      ],
25	      "label": null,
26	      "suggested_replacement": null,
+	      "suggestion_approximate": null,
27	      "expansion": null
28	    }
29	  ],

51	          ],
52	          "label": null,
53	          "suggested_replacement": null,
+	          "suggestion_approximate": null,
54	          "expansion": null
55	        }
56	      ],

80	          ],
81	          "label": null,
82	          "suggested_replacement": "1 / (2 + 3)",
+	          "suggestion_approximate": null,
83	          "expansion": null
84	        }
85	      ],


The actual stderr differed from the expected stderr.
Actual stderr saved to /nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui/lint/unused_parens_json_suggestion.stderr
To update references, run this command from build directory:
/nobackup/rust/src/test/ui/update-references.sh '/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui' 'lint/unused_parens_json_suggestion.rs'

error: 1 errors occurred comparing output.
status: exit code: 0
command: "/nobackup/rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/nobackup/rust/src/test/ui/lint/unused_parens_json_suggestion.rs" "-L" "/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui/lint/unused_parens_json_suggestion.stage1-x86_64-unknown-linux-gnu" "-Crpath" "-O" "-Zmiri" "-Zunstable-options" "-Lnative=/nobackup/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--error-format" "pretty-json" "-Zunstable-options" "-L" "/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui/lint/unused_parens_json_suggestion.stage1-x86_64-unknown-linux-gnu.aux" "-A" "unused"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
{
  "message": "unnecessary parentheses around assigned value",
  "code": {
    "code": "unused_parens",
    "explanation": null
  },
  "level": "warning",
  "spans": [
    {
      "file_name": "/nobackup/rust/src/test/ui/lint/unused_parens_json_suggestion.rs",
      "byte_start": 1056,
      "byte_end": 1069,
      "line_start": 25,
      "line_end": 25,
      "column_start": 14,
      "column_end": 27,
      "is_primary": true,
      "text": [
        {
          "text": "    let _a = (1 / (2 + 3));",
          "highlight_start": 14,
          "highlight_end": 27
        }
      ],
      "label": null,
      "suggested_replacement": null,
      "suggestion_approximate": null,
      "expansion": null
    }
  ],
  "children": [
    {
      "message": "lint level defined here",
      "code": null,
      "level": "note",
      "spans": [
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/unused_parens_json_suggestion.rs",
          "byte_start": 902,
          "byte_end": 915,
          "line_start": 20,
          "line_end": 20,
          "column_start": 9,
          "column_end": 22,
          "is_primary": true,
          "text": [
            {
              "text": "#![warn(unused_parens)]",
              "highlight_start": 9,
              "highlight_end": 22
            }
          ],
          "label": null,
          "suggested_replacement": null,
          "suggestion_approximate": null,
          "expansion": null
        }
      ],
      "children": [],
      "rendered": null
    },
    {
      "message": "remove these parentheses",
      "code": null,
      "level": "help",
      "spans": [
        {
          "file_name": "/nobackup/rust/src/test/ui/lint/unused_parens_json_suggestion.rs",
          "byte_start": 1056,
          "byte_end": 1069,
          "line_start": 25,
          "line_end": 25,
          "column_start": 14,
          "column_end": 27,
          "is_primary": true,
          "text": [
            {
              "text": "    let _a = (1 / (2 + 3));",
              "highlight_start": 14,
              "highlight_end": 27
            }
          ],
          "label": null,
          "suggested_replacement": "1 / (2 + 3)",
          "suggestion_approximate": null,
          "expansion": null
        }
      ],
      "children": [],
      "rendered": null
    }
  ],
  "rendered": "warning: unnecessary parentheses around assigned value\n  --> /nobackup/rust/src/test/ui/lint/unused_parens_json_suggestion.rs:25:14\n   |\n25 |     let _a = (1 / (2 + 3));\n   |              ^^^^^^^^^^^^^ help: remove these parentheses\n   |\nnote: lint level defined here\n  --> /nobackup/rust/src/test/ui/lint/unused_parens_json_suggestion.rs:20:9\n   |\n20 | #![warn(unused_parens)]\n   |         ^^^^^^^^^^^^^\n\n"
}

------------------------------------------

thread '[ui] ui/lint/unused_parens_json_suggestion.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:2892:9


failures:
    [ui] ui/explain.rs
    [ui] ui/lint/unused_parens_json_suggestion.rs
    [ui] ui/lint/use_suggestion_json.rs

test result: FAILED. 0 passed; 3 failed; 1233 ignored; 0 measured; 0 filtered out

thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:476:22


command did not execute successfully: "/nobackup/rust/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/nobackup/rust/build/x86_64-unknown-linux-gnu/stage1/lib" "--run-lib-path" "/nobackup/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/nobackup/rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "--src-base" "/nobackup/rust/src/test/ui" "--build-base" "/nobackup/rust/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage1-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/nobackup/rust/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zmiri -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zmiri -Zunstable-options  -Lnative=/nobackup/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python" "--lldb-python" "/usr/bin/python" "--gdb" "/usr/bin/gdb" "--llvm-version" "6.0.0\n" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" ""
expected success, got: exit code: 101


failed to run: /nobackup/rust/build/bootstrap/debug/bootstrap test --stage 1 src/test/ui src/test/compile-fail src/test/run-pass
Build completed unsuccessfully in 0:00:06
```
</details>

**Comment by mark-i-m** [comment]
They seem completely unrelated, which I find confusing...

File: src/librustc_resolve/diagnostics.rs (hunk 2 above)
Comment (mark-i-m) [reply src/librustc_resolve/diagnostics.rs]: Hmm... maybe if I run the full test suite? Let me try.

Comment (mark-i-m) [reply src/librustc_resolve/diagnostics.rs]: Yep, it appears that tidy does not run for a full build...

Comment (mark-i-m) [reply src/librustc_resolve/diagnostics.rs]: Ok, I changed it to E0909

**Comment by estebank** [comment]
@mark-i-m oh, those. I believe that with a fresh clone of the repo those error go away... I've just been ignoring them :-/

If those are the only ones failing, travis will probably succeed.

**Comment by mark-i-m** [comment]
That's good to know... maybe it's because I'm running tests against the stage1 compiler locally?

File: src/librustc_resolve/lib.rs [hunk 3]
```diff
@@ -395,12 +395,19 @@ enum AliasPossibility {
...
+    Bound,
+}
+
 #[derive(Copy, Clone, PartialEq, Eq, Debug)]
 enum PathSource<'a> {
     // Type paths `Path`.
     Type,
     // Trait paths in bounds or impls.
-    Trait(AliasPossibility),
+    Trait(AliasPossibility, ImplOrBound),
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: Wait.
`AliasPossibility` is supposed to be exactly the same thing as `ImplOrBound` - aliases can be used in bounds, but cannot be used in `impl Trait for ...`.
Also, poly-trait-ref is *always* a bound, no need to change visitors.
What this PR tries to achieve?

Comment (mark-i-m) [reply src/librustc_resolve/lib.rs]: Thanks @petrochenkov ! That's really useful info. The goal was simply to emit a different error if something that is not a trait is used in a bound as opposed to an impl (i.e. `impl Struct for Foo` vs `foo<T: Struct>(...) {}`).

Comment (mark-i-m) [reply src/librustc_resolve/lib.rs]: So you are saying that if `AliasPossibility::Maybe` it _must_ be a bound. But is it also true that if `AliasPossibility::No` it _must_ be an `impl`?

Comment (mark-i-m) [reply src/librustc_resolve/lib.rs]: @petrochenkov Is this correct:

```diff
diff --git a/src/libsyntax/visit.rs b/src/libsyntax/visit.rs
index bf8beb1..5492cd6 100644
--- a/src/libsyntax/visit.rs
+++ b/src/libsyntax/visit.rs
@@ -548,9 +541,9 @@ impl<'a> PathSource<'a> {
         __diagnostic_used!(E0578);
         __diagnostic_used!(E0909);
         match (self, has_unexpected_resolution) {
-            (PathSource::Trait(_, ImplOrBound::Impl), true) => "E0404",
-            (PathSource::Trait(_, ImplOrBound::Bound), true) => "E0909",
-            (PathSource::Trait(_, _), false) => "E0405",
+            (PathSource::Trait(AliasPossibility::No), true) => "E0404",
+            (PathSource::Trait(AliasPossibility::Maybe), true) => "E0909",
+            (PathSource::Trait(_), false) => "E0405",
             (PathSource::Type, true) => "E0573",
             (PathSource::Type, false) => "E0412",
             (PathSource::Struct, true) => "E0574",
```

Comment (petrochenkov) [reply src/librustc_resolve/lib.rs]: No, the existing code is already correct, only `E0245` need to be removed as dead code.
Also docs for `E0404` should probably be tweaked to mention bounds in addition to impls.

Comment (petrochenkov) [reply src/librustc_resolve/lib.rs]: To clarify, https://github.com/rust-lang/rust/pull/48446#discussion_r170138387 is correct too, if you really want to make these error codes more fine grained, I'm just not sure this makes sense given that trait aliases are not even implemented yet.

If you do want to spend time on this, I'd suggest to split `E0405` as well and document differences between bounds and impls and how trait aliases can be used in bounds but not in impls in the descriptions of `E0404`/`E0405`/`E0909`/`E0910`.

Comment (mark-i-m) [reply src/librustc_resolve/lib.rs]: >  if you really want to make these error codes more fine grained

I think that was the conclusion of #36337

> If you do want to spend time on this, I'd suggest to split E0405 as well and document differences between bounds and impls and how trait aliases can be used in bounds but not in impls in the descriptions of E0404/E0405/E0909/E0910.

Sure, I can try to do that.

**Comment by mark-i-m** [comment]
Made this WIP again... I will look into splitting E0405 too. In the meantime, I removed the `ImplOrBound` enum.

**Comment by mark-i-m** [comment]
@petrochenkov

> If you do want to spend time on this, I'd suggest to split E0405 as well and document differences between bounds and impls and how trait aliases can be used in bounds but not in impls in the descriptions of E0404/E0405/E0909/E0910.

Hmm... Reading this again, I'm not 100% sure we need to mention trait aliases at all. Can't we just mention say that structs are not allowed as bounds?

**Comment by petrochenkov** [comment]
That's what I suggested originally. If trait aliases are not mentioned, then there's no difference between impls and bounds, structs are not allowed "in trait position" generally (i.e. in both impls and bounds), and splitting error codes is not necessary, only error descriptions need to be updated.

**Comment by mark-i-m** [comment]
@petrochenkov @estebank Updated.
- Remove E0245
- Remove E0909 (or rather, I didn't add any new error codes)
- Expanded the explanation of E0404

Let me know what you think.

**Comment by estebank** [comment]
LGTM.

r? @petrochenkov

**Comment by petrochenkov** [comment]
Thanks!
r=me after squashing commits and updating the failing test

**Comment by mark-i-m** [comment]
@bors r+

:tada:

**Comment by mark-i-m** [comment]
Oh, I should have read "after squashing commits"!

**Comment by mark-i-m** [comment]
Ok, I squashed and fixed the test. Will wait for travis, then r+

**Pull request merged**